pkg/operators/wasm/wasm.go

```diff
@@ -0,0 +1,235 @@
+) (
+	operators.ImageOperatorInstance, error,
+) {
+	return &wasmOperatorInstance{
+		desc:      desc,
+		gadgetCtx: gadgetCtx,
+		handleMap: map[uint32]any{},
+		logger:    gadgetCtx.Logger(),
+	}, nil
+}
+
+type wasmOperatorInstance struct {
+	desc      ocispec.Descriptor
+	rt        wazero.Runtime
+	gadgetCtx operators.GadgetContext
+	mod       wapi.Module
+
+	logger logger.Logger
+
+	// malloc function exported by the guest
+	guestMalloc wapi.Function
+
+	// Golang objects are exposed to the wasm module by using a handleID
+	handleMap  map[uint32]any
+	handleCtr  uint32
+	handleLock sync.RWMutex
+}
+
+func (i *wasmOperatorInstance) Name() string {
+	return "wasm"
+}
+
+func (i *wasmOperatorInstance) Prepare(gadgetCtx operators.GadgetContext) error {
+	err := i.init(gadgetCtx)
+	if err != nil {
+		return fmt.Errorf("initializing wasm: %w", err)
+	}
+	err = i.Init(gadgetCtx)
+	if err != nil {
+		return fmt.Errorf("initializing wasm guest: %w", err)
+	}
+
+	return nil
+}
+
+func (i *wasmOperatorInstance) ExtraParams(gadgetCtx operators.GadgetContext) api.Params {
+	return nil
+}
+
+func (i *wasmOperatorInstance) addHandle(obj any) uint32 {
+	if obj == nil {
+		return 0
+	}
+
+	i.handleLock.Lock()
+	defer i.handleLock.Unlock()
+	i.handleCtr++
+	if i.handleCtr == 0 { // 0 is reserved
+		i.handleCtr++
+	}
+	xctr := 0
+	for {
+		if xctr > 1<<32 {
+			// exhausted; TODO: report somehow
+			return 0
+		}
+		if _, ok := i.handleMap[i.handleCtr]; !ok {
+			// register new entry
+			i.handleMap[i.handleCtr] = obj
+			return i.handleCtr
+		}
+		xctr++
+	}
+}
+
+func (i *wasmOperatorInstance) getHandle(handleID uint32) any {
+	i.handleLock.RLock()
+	defer i.handleLock.RUnlock()
+	return i.handleMap[handleID]
+}
+
+func (i *wasmOperatorInstance) delHandle(handleID uint32) {
+	i.handleLock.Lock()
+	defer i.handleLock.Unlock()
+	delete(i.handleMap, handleID)
+}
+
+func (i *wasmOperatorInstance) init(gadgetCtx operators.GadgetContext) error {
+	ctx := gadgetCtx.Context()
+	rtConfig := wazero.NewRuntimeConfig().
+		WithCloseOnContextDone(true).
+		WithMemoryLimitPages(256) // 16MB (64KB per page)
+	i.rt = wazero.NewRuntimeWithConfig(ctx, rtConfig)
+
+	env := i.rt.NewHostModuleBuilder("env")
+
+	i.addLogFuncs(env)
+	i.addDataSourceFuncs(env)
+	i.addFieldFuncs(env)
+
+	if _, err := env.Instantiate(ctx); err != nil {
+		return fmt.Errorf("instantiating host module: %w", err)
+	}
+
+	wasi_snapshot_preview1.MustInstantiate(ctx, i.rt)
+
+	reader, err := oci.GetContentFromDescriptor(gadgetCtx.Context(), i.desc)
+	if err != nil {
+		return fmt.Errorf("getting wasm program: %w", err)
+	}
+
+	buf := new(bytes.Buffer)
+	buf.ReadFrom(reader)
+	reader.Close()
+
+	config := wazero.NewModuleConfig()
+	mod, err := i.rt.InstantiateWithConfig(ctx, buf.Bytes(), config)
+	if err != nil {
+		return fmt.Errorf("instantiating wasm: %w", err)
+	}
+	i.mod = mod
+
+	// We need to call malloc on the guest to pass strings
+	i.guestMalloc = mod.ExportedFunction("malloc")
```

Comment:
Needs a check, right?

Reply:
Good catch. Yes, we need malloc to handle bytes and strings. We'll need to thing how it works with other languages, according to the wazero examples [malloc is exported by tinygo](https://github.com/tetratelabs/wazero/blob/5a8a053bff0ae795b264de9672016745cb842070/examples/allocation/tinygo/greet.go#L55) by I see in rust the user needs to provide a pair of [allocate / deallocate](https://github.com/tetratelabs/wazero/blob/5a8a053bff0ae795b264de9672016745cb842070/examples/allocation/rust/greet.go#L50-L51) functions.

I'll keep using malloc (and erroring out if it's not exported) and open an issue to track this down.

Reply:
Created #3073